xpt2046/xpt2046.go

```diff
@@ -0,0 +1,175 @@
+// Package xpt2046 implements a driver for the XPT2046 resistive touch controller as packaged on the TFT_320QVT board
+//
+// Datasheet: http://grobotronics.com/images/datasheets/xpt2046-datasheed.pdf
+package xpt2046
+
+import (
+	"machine"
+	"time"
+
+	"tinygo.org/x/drivers/touch"
+)
+
+type Device struct {
+	t_clk  machine.Pin
+	t_cs   machine.Pin
+	t_din  machine.Pin
+	t_dout machine.Pin
+	t_irq  machine.Pin
+
+	precision uint8
+}
+
+func New(t_clk, t_cs, t_din, t_dout, t_irq machine.Pin, precision uint8) Device {
+	return Device{
+		precision: precision,
+
+		t_clk:  t_clk,
+		t_cs:   t_cs,
+		t_din:  t_din,
+		t_dout: t_dout,
+		t_irq:  t_irq,
+	}
+}
+
+func (d *Device) Configure() {
+	d.t_clk.Configure(machine.PinConfig{Mode: machine.PinOutput})
+	d.t_cs.Configure(machine.PinConfig{Mode: machine.PinOutput})
+	d.t_din.Configure(machine.PinConfig{Mode: machine.PinOutput})
+
+	d.t_dout.Configure(machine.PinConfig{Mode: machine.PinInput})
+	d.t_irq.Configure(machine.PinConfig{Mode: machine.PinInput})
+
+	d.t_clk.Low()
+	d.t_cs.High()
+	d.t_din.Low()
+
+	d.readRaw() //Set Powerdown mode to enable T_IRQ
+}
+
+func busSleep() {
+	time.Sleep(5 * time.Nanosecond)
+}
+
+func pulseHigh(p machine.Pin) {
+	p.High()
+	busSleep()
+	p.Low()
+	busSleep()
+}
+
+func (d *Device) writeCommand(data uint8) {
+
+	for count := uint8(0); count < 8; count++ {
+		d.t_din.Set((data & 0x80) != 0)
+		data <<= 1
+		pulseHigh(d.t_clk)
+	}
+
+}
+
+func (d *Device) readData() uint16 {
+
+	data := uint16(0)
+
+	for count := uint8(0); count < 12; count++ {
+		data <<= 1
+		pulseHigh(d.t_clk)
+		if d.t_dout.Get() {
+			data |= 1
+		}
+	}
+	pulseHigh(d.t_clk) //13
+	pulseHigh(d.t_clk) //14
+	pulseHigh(d.t_clk) //15
+	pulseHigh(d.t_clk) //16
+
+	return data
+}
+
+func (d *Device) ReadTouchPoint() touch.Point {
+
+	tx := uint32(0)
+	ty := uint32(0)
+	tz := uint32(0)
+	sampleCount := uint8(0)
+
+	d.t_cs.Low()
+
+	for ; sampleCount < d.precision && d.Touched(); sampleCount++ {
```

Comment:
If d.precision is 0, `readRaw()` will not be executed.
The initial value of d.precision should be a value greater than or equal to 1.

Reply:
I set a default of 10 if 0 is passed.